### ISSUE TITLE
Pulling images from private registry through harbor cache-proxy

### ISSUE THREAD

**Issue by pavanfhw** [k3s-io/k3s#2504]
**Environmental Info:**
K3s Version: 

k3s version v1.18.6+k3s1 (6f56fa1d)

Node(s) CPU architecture, OS, and Version: 

Linux <hostname> 5.4.0-37-generic

Cluster Configuration:

1 master, 5 workers
**Describe the bug:**

I'm having problems pulling images from my private registry using a Harbor registry configured as a cache-proxy. In the description I will call my private registry **target registry** and the Harbor proxy-cache **proxy registry**
I have tried 2 solutions:

First:
Just pull directly from the proxy registry. Harbor proxy-cache should be able to respond pull request to all images present on the target registry. But `crictl` is unable to do so if the image was not successfully pulled previously. `dockel pull` works for example.
`crictl pull` only works with images already 'cached' on the proxy registry. Example

`k3s crictl pull <proxy-registry><proxy-cache-project><image>:1.0` **fails** `image not found`
`docker pull <proxy-registry><proxy-cache-project><image>:1.0` (from another machine outside the cluster) **success**
`k3s crictl pull <proxy-registry><proxy-cache-project><image>:1.0` (again) **success** this time
`k3s crictl pull <proxy-registry><proxy-cache-project><image>:1.1` **fails**

Second:
Use `/etc/rancher/k3s/registries.yaml` (https://rancher.com/docs/k3s/latest/en/installation/private-registry/). Mirror configuration:

```
mirrors:
  <target registry>:
    endpoint:
      - "<proxy registry>"
configs:
  "<proxy registry>":
    auth:
      username: *****           # proxy registry credentials
      password: ***********   # are this the right credentials to use here or should they be from the private registry?
    tls:
      cert_file: /home/rancher/registry/cert/fullchain.pem    # proxy registry certificate, key and ca certificate
      key_file: /home/rancher/registry/cert/privkey.pem      # are these the right files to use here or should they be 
      ca_file: /home/rancher/registry/cert/chain.pem          # from the private registry?
```

With this approach running 

`cri pull <target registry><image>:<tag>`

I get `unexpected status code [manifests 0.23.0]: 401 Unauthorized`

Any help for any of the two approachs

**Steps To Reproduce:**

- Create k3s cluster
- Create Harbor proxy-cache registry (https://goharbor.io/docs/2.1.0/administration/configure-proxy-cache/)
- Try to pull images using `crictl`
- Confirm that `docker pull` works for any image in the target registry
- Confirm that `crictl pull` only works for previously pulled images

**Expected behavior:**

`k3s crictl pull` be able to pull any image present on the target registry
**Actual behavior:**

`k3s crictl pull` only works if someone pulled the image previously. `crictl` does not trigger the Harbor registry to pull the image from the target registry if it does not have it.

**Additional context / logs:**

**Comment by brandond** [comment]
Have you reported this against either harbor or containerd?

**Comment by pavanfhw** [comment]
Yes I did, in both. I am also trying here in case anyone with a successful experience could help. And to confirm if it happened to other k3s users.

**Comment by brandond** [comment]
Would you mind linking the issues? Also, you might try with 1.19, as it uses a much newer version of containerd (1.40).  An upcoming release will include 1.4.1 which includes a number of bugfixes.

**Comment by pavanfhw** [comment]
Similar issues opened in Containerd and Harbor
[Containerd #4706](https://github.com/containerd/containerd/issues/4706)
[Harbor #13475](https://github.com/goharbor/harbor/issues/13475)

**Comment by pavanfhw** [comment]
Ok, will try k3s 1.19

**Comment by alxlenc** [comment]
Which Harbor version are you using? There was an [issue](https://github.com/goharbor/harbor/issues/13153) fixed in 2.1.1 which was supposed to fix fetching non cached images using containerd. 

I have not tried again since it was released and reverted to use k3s with dockerhub and Harbor only for my own images. 

This was my `registries.yaml` which successfully fetched pre-cached images:

```
    mirrors:
      docker.io:
        endpoint:
          - "https://$IMAGEREG_HOST/v2/$HARBOR_PROXY_PROJECT"
      $IMAGEREG_HOST:
        endpoint:
          - "https://$IMAGEREG_HOST"
        
    configs:
      "$IMAGEREG_HOST":
        auth:
          username: $IMAGEREG_USSER
          password: $IMAGEREG_PASSWORD
        tls:
          cert_file: /etc/node/ci-cd.crt
          key_file: /etc/node/ci-cd.key
          ca_file: /etc/ssl/certs/ca-certificates.crt
```

**Comment by pavanfhw** [comment]
Updating: resutls using k3 1.19 were the same. I resolved the issue for now using a proxy that has access to the target registry.
I will try later upgrading harbor version to 2.1.1.

**Comment by brandond** [comment]
That harbor issue looks like exactly what you're running into.

**Comment by brandond** [comment]
Any update, or can we close this?

**Comment by alxlenc** [comment]
In my case I have updated Harbor to 2.1.2 and the pull through works ok.

**Comment by brandond** [comment]
Thanks for the update!

**Comment by ricardojdsilva87** [comment]
> Which Harbor version are you using? There was an [issue](https://github.com/goharbor/harbor/issues/13153) fixed in 2.1.1 which was supposed to fix fetching non cached images using containerd.
> 
> I have not tried again since it was released and reverted to use k3s with dockerhub and Harbor only for my own images.
> 
> This was my `registries.yaml` which successfully fetched pre-cached images:
> 
> ```
>     mirrors:
>       docker.io:
>         endpoint:
>           - "https://$IMAGEREG_HOST/v2/$HARBOR_PROXY_PROJECT"
>       $IMAGEREG_HOST:
>         endpoint:
>           - "https://$IMAGEREG_HOST"
>         
>     configs:
>       "$IMAGEREG_HOST":
>         auth:
>           username: $IMAGEREG_USSER
>           password: $IMAGEREG_PASSWORD
>         tls:
>           cert_file: /etc/node/ci-cd.crt
>           key_file: /etc/node/ci-cd.key
>           ca_file: /etc/ssl/certs/ca-certificates.crt
> ```

Hello all,
This extra setting works perfectly for `crictl` and `containerd`. We are using kops with a kubernetes version 1.19.13 with the containerd as the engine.

Thanks!
